src/meltano/migrations/versions/f4c225a9492f_create_dedicated_job_state_table.py

```diff
@@ -0,0 +1,404 @@
+            .order_by(Job.ended_at.desc())
+            .first()
+        )
+
+    @classmethod
+    def all_stale(cls, session):
+        """Return all stale states.
+
+        Args:
+            session: the session to use to query the db
+
+        Returns:
+            All stale states with any state ID
+        """
+        now = datetime.utcnow()
+        last_valid_heartbeat_at = now - timedelta(minutes=HEARTBEAT_VALID_MINUTES)
+        last_valid_started_at = now - timedelta(hours=HEARTBEATLESS_JOB_VALID_HOURS)
+
+        return session.query(Job).filter(
+            (Job.state == State.RUNNING)  # noqa: WPS465
+            & (
+                (
+                    Job.last_heartbeat_at.isnot(None)  # noqa: WPS465
+                    & (Job.last_heartbeat_at < last_valid_heartbeat_at)
+                )
+                | (
+                    Job.last_heartbeat_at.is_(None)  # noqa: WPS465
+                    & (Job.started_at < last_valid_started_at)
+                )
+            )
+        )
+
+    def stale(self, session):
+        """Return stale states with the instance's state ID.
+
+        Args:
+            session: the session to use in querying the db
+
+        Returns:
+            All stale states with instance's state ID
+        """
+        return self.all_stale(session).filter(Job.job_name == self.state_id)
+
+    def get_all(self, session: object, since=None):
+        """Return all state with the instance's state ID.
+
+        Args:
+            session: the session to use in querying the db
+            since: only return state which ended after this datetime
+
+        Returns:
+            All state with instance's state ID which ended after 'since'
+        """
+        query = (
+            session.query(Job)
+            .filter(Job.job_name == self.state_id)
+            .order_by(Job.ended_at.asc())
+        )
+
+        if since:
+            query = query.filter(Job.ended_at > since)
+        return query
+
+
+# Copied from core/job/job.py
+HEARTBEATLESS_JOB_VALID_HOURS = 24
+HEARTBEAT_VALID_MINUTES = 5
+
+
+class State(Enum):
+    """Represents status of a Job."""
+
+    IDLE = (0, ("RUNNING", "FAIL"))
+    RUNNING = (1, ("SUCCESS", "FAIL"))
+    SUCCESS = (2, ())
+    FAIL = (3, ("RUNNING",))
+    DEAD = (4, ())
+    STATE_EDIT = (5, ())
+
+
+class Payload(IntFlag):
+    """Flag indicating whether a Job has state in its payload field."""
+
+    STATE = 1
+    INCOMPLETE_STATE = 2
+
+
+class Job(SystemModel):  # noqa: WPS214
+    """Model class that represents a `meltano elt` run in the system database.
+
+    Includes State.STATE_EDIT rows which represent CLI invocations of the
+    `meltano state` command which wrote state to the db. Queries that are
+    meant to return only actual job runs should filter out records with
+    state == State.STATE_EDIT.
+    """
+
+    __tablename__ = "runs"
+
+    id = Column(types.Integer, primary_key=True)
+    job_name = Column(types.String)
```

Comment:
qq - do we need to throw some max lengths on these String and JSONEncodedDict columns ? To maintain compatability with MSSQL (and MySQL later).

basically thinking of:

- #6167

Reply:
Yeah we definitely do need max length down in the `JobState` table. I've added that for the `JSONEncodedDict`. Turns out MSSQL also requires primary key columns to be <=900 bytes as well, so I've added that constraint there to.